pkg/controller/topology_test.go

```diff
@@ -1493,6 +1494,251 @@ func TestPreferredTopologies(t *testing.T) {
 	}
 }
 
+func TestProvisionWithDeletedNodeFromCache(t *testing.T) {
+	nodeLabels := []map[string]string{
+		{"com.example.csi/zone": "zone1", "com.example.csi/rack": "rackA"},
+		{"com.example.csi/zone": "zone2", "com.example.csi/rack": "rackB"},
+	}
+	topologyKeys := []map[string][]string{
+		{testDriverName: {"com.example.csi/zone", "com.example.csi/rack"}},
+		{testDriverName: {"com.example.csi/zone", "com.example.csi/rack"}},
+	}
+	pvcUID := "a9e2d5a3-1c64-4787-97b7-1a22d5a0b123"
+	pvcName := "testpvc"
+
+	testcases := []struct {
+		name              string
+		strictTopology    bool
+		immediateTopology bool
+		allowedTopologies []v1.TopologySelectorTerm
+		selectedNode      string
+		expectedRequisite []*csi.Topology
+		expectedPreferred []*csi.Topology
+	}{
+		{
+			name:              "delayed-binding, non-strict, without-allowed-topology",
+			strictTopology:    false,
+			immediateTopology: false,
+			allowedTopologies: nil,
+			selectedNode:      "node-0",
+			expectedRequisite: []*csi.Topology{
+				{Segments: map[string]string{"com.example.csi/zone": "zone1", "com.example.csi/rack": "rackA"}},
+				{Segments: map[string]string{"com.example.csi/zone": "zone2", "com.example.csi/rack": "rackB"}},
+			},
+			expectedPreferred: []*csi.Topology{
+				{Segments: map[string]string{"com.example.csi/zone": "zone1", "com.example.csi/rack": "rackA"}},
+				{Segments: map[string]string{"com.example.csi/zone": "zone2", "com.example.csi/rack": "rackB"}},
+			},
+		},
+		{
+			name:              "delayed-binding, non-strict, with-allowed-topology",
+			strictTopology:    false,
+			immediateTopology: false,
+			allowedTopologies: []v1.TopologySelectorTerm{
+				{
+					MatchLabelExpressions: []v1.TopologySelectorLabelRequirement{
+						{Key: "com.example.csi/zone", Values: []string{"zone1"}},
```

Comment:
for the allowed topology test cases, it would be better to use values that are different from what we store in the cache so that we can be sure we're not using the cache values.

Reply:
But this is to test the node has been deleted? How can we pass in a different set of allowedTopologies if it is for the same CreateVolume retry?

Reply:
For 2, I restrict to 2 because i was intentionally deleting the first node to make the node "deleted".